### PR TITLE
Avoid exception when CRD doesn't follow Pod spec shape

### LINKED ISSUE #1194: Avoid exception when CRD doesn't follow Pod spec shape
## Expected Behavior
KEDA should be able to scale custom resources as long as the target CR defines /scale subresource

## Actual Behavior
For [SeldonDeployment CR ](https://github.com/SeldonIO/seldon-core/blob/master/operator/apis/machinelearning.seldon.io/v1/seldondeployment_types.go) 

Got the following exception 

```
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:90
2020-09-24T02:52:10.360Z        ERROR   controller      Reconciler error        {"reconcilerGroup": "keda.sh", "reconcilerKind": "ScaledObject", "controller": "scaledobject", "name": "agao-test-keda", "namespace": "default",
"error": "no containers found"}
github.com/go-logr/zapr.(*zapLogger).Error
        /go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:237
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:209
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.2/pkg/internal/controller/controller.go:188
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:90
```

It seems there is a requirement for the shape the CR as well https://github.com/kedacore/keda/blob/v2/pkg/scaling/scale_handler.go#L349 ? 

## Steps to Reproduce the Problem

  1. Install KEDA
  2. Deploy a ScaledObject
```
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: agao-test-keda
spec:
  scaleTargetRef:
    apiVersion: machinelearning.seldon.io/v1
    kind: SeldonDeployment
    name: agao-test
  pollingInterval: 15                                # Optional. Default: 30 seconds
  cooldownPeriod:  30                               # Optional. Default: 300 seconds
  minReplicaCount: 0                                 # Optional. Default: 0
  maxReplicaCount: 100                               # Optional. Default: 100
  triggers:
  - type: prometheus
    metadata:
      serverAddress: http://prometheus.istio-system.svc.cluster.local:9090
      metricName: access_frequency
      threshold: '3'
```
  3. View operator logs

## Specifications

  - **KEDA Version:** v2.0.0-beta
  - **Kubernetes Version:** v1.17.5
  - **Scaler(s):** Prometheus

### PR DESCRIPTION
**Pull request by anggao**
Signed-off-by: Ang Gao <ang.gao87@gmail.com>

Patches to avoid exception when CRD doesn't follow Pod spec shape

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)

Fixes #1194

### CODE REVIEW

File: pkg/scaling/scale_handler.go [hunk 1]
```diff
@@ -296,15 +296,21 @@ func min(x, y int64) int64 {
...
+		}
 	}
 
 	for i, trigger := range withTriggers.Spec.Triggers {
-		authParams, podIdentity := resolver.ResolveAuthRef(h.client, logger, trigger.AuthenticationRef, &podTemplateSpec.Spec, withTriggers.Namespace)
-
+		var authParams map[string]string
+		var podIdentity string
+		if podTemplateSpec != nil {
+			authParams, podIdentity = resolver.ResolveAuthRef(h.client, logger, trigger.AuthenticationRef, &podTemplateSpec.Spec, withTriggers.Namespace)
```
Comment (zroubalik) [inline pkg/scaling/scale_handler.go:312]: We shouldn't skip the whole `resolver.ResolveAuthRef()` call, if you look inside, it is resolving env vars from TriggerAuthentication resource as well (which should be ok even if you don't have podSpec).

File: pkg/scaling/scale_handler.go [hunk 2]
```diff
@@ -348,7 +354,8 @@ func (h *scaleHandler) getPods(scalableObject interface{}) (*corev1.PodTemplateS
 
 		if withPods.Spec.Template.Spec.Containers == nil {
 			h.logger.Info("There aren't any containers in the ScaleTarget", "resource", obj.Status.ScaleTargetGVKR.GVKString(), "name", obj.Spec.ScaleTargetRef.Name)
-			return nil, "", fmt.Errorf("no containers found")
+			h.logger.Info("You may not be able to use all scaler types")
```
Comment (zroubalik) [inline pkg/scaling/scale_handler.go:357]: No need to add a second log message, we should update the one above to message like this:

`There aren't any containers found in the ScaleTarget, therefore it is no possible to inject environment properties`

**Review by zroubalik — CHANGES_REQUESTED** [review]

File: pkg/scaling/scale_handler.go (hunk 2 above)
Comment (zroubalik) [reply pkg/scaling/scale_handler.go:357]: And this log message should be probably debug one: `h.logger.V(1).Info(...)`

Comment (anggao) [reply pkg/scaling/scale_handler.go]: Yeah that's better fixed.

File: pkg/scaling/scale_handler.go (hunk 1 above)
Comment (anggao) [reply pkg/scaling/scale_handler.go]: I see, I added it back. but within `resolver.ResolveAuthRef()` the calll to `ResolveContainerEnv` assume `podSpec` is not `nil` thus I added a condition to it

Comment (anggao) [reply pkg/scaling/scale_handler.go]: @zroubalik let me know whether there are more changes I should add

**Review by zroubalik — APPROVED** [review]
LGTM

**Review by ahmelsayed — APPROVED** [review]

**Pull request merged**